### ISSUE TITLE
Enhancement: Possibility to unregister route handler 

### ISSUE THREAD

**Issue by pjnssn** [starlite-api/starlite#1276]
There is a way to dynamically register routes with `request.app.register(handler)`, it would be nice to have a way to dynamically unregister these routes.

**Comment by provinzkraut** [comment]
What would the use case be for unregistering a route handler?

**Comment by JacobCoffee** [comment]
I may be incorrect but I believe this is a followup from a Discord discussion

https://discord.com/channels/919193495116337154/1081644750853116036

**Comment by Goldziher** [comment]
That's from discord, yes. 


So, I'm not sure about this one - it's worthy if a discussion. 

Unregistering routes is not very simple to implement - probably. I might be wrong here, but we will need to do cleanup in several places to achieve this. It's still perfectly doable if we want to implement - we need to make a decision on this though. 

The way to "fix" this on you end, i would say, is to use some custom Guard function in combination with the application state. It will allow you to dynamically gate routes- but is not a builtin solution. 

I don't see a problem with implementing this, so let's see what @provinzkraut and maybe others have to say.

**Comment by provinzkraut** [comment]
I think I would like to see the actual use case and reasoning for why this has to be implemented this way, because I feel like something along the lines of what @Goldziher suggested would be a more proper solution and I can't imagine a case where dynamically unregistering a route handler is the best way to go.

**Comment by pjnssn** [comment]
The use-case for this would be for a REST API mocking service, something along the lines of [wiremock](https://wiremock.org/). This API would be constantly running and could be interacted with by a user. Something like: a user would have a test case where they define a set of endpoints to mock, they POST some endpoint definitions to lets say /register which is responsible for creating this endpoint to the spec provided. Then once the test is done, you call lets say /unregister and it removes the endpoint from being used.

I had a look at some of the source code and as @Goldziher said I see how it could be a fair bit of work to implement as there are places where you use a caching for example. I can definitely see how unregistering endpoints is very much an edge case need, and while I think for this use case it would be a cleaner solution to have unregistering available, I definitely see how a similar effect can be achieved in other ways.

**Comment by provinzkraut** [comment]
> a user would have a test case where they define a set of endpoints to mock, they POST some endpoint definitions to lets say /register which is responsible for creating this endpoint to the spec provided. Then once the test is done, you call lets say /unregister and it removes the endpoint from being used.

I would say the proper way to do this is to dynamically create an application based on the spec, run that and expose it to the user, and discard it when they're done.
